src/Microsoft.ML.Transforms/CustomMappingCatalog.cs

```diff
@@ -46,5 +46,64 @@ public static class CustomMappingCatalog
+        /// If the resulting transformer needs to be save-able, the class defining <paramref name="mapAction"/> should implement
+        /// <see cref="StatefulCustomMappingFactory{TSrc, TState, TDst}"/> and needs to be decorated with
+        /// <see cref="CustomMappingFactoryAttributeAttribute"/> with the provided <paramref name="contractName"/>.
+        /// The assembly containing the class should be registered in the environment where it is loaded back
+        /// using <see cref="ComponentCatalog.RegisterAssembly(System.Reflection.Assembly, bool)"/>.</param>
+        /// <param name="stateInitAction">The action to initialize the state object, that is called once before the cursor is initialized.</param>
+        /// <param name="contractName">The contract name, used by ML.NET for loading the model.
+        /// If <see langword="null"/> is specified, resulting transformer would not be save-able.</param>
+        /// <param name="inputSchemaDefinition">Additional parameters for schema mapping between <typeparamref name="TSrc"/> and input data.
+        /// Useful when dealing with annotations.</param>
+        /// <param name="outputSchemaDefinition">Additional parameters for schema mapping between <typeparamref name="TDst"/> and output data.
+        /// Useful when dealing with annotations.</param>
+        public static StatefulCustomMappingEstimator<TSrc, TState, TDst> StatefulCustomMapping<TSrc, TState, TDst>(this TransformsCatalog catalog, Action<TSrc, TState, TDst> mapAction,
+            Action<TState> stateInitAction, string contractName, SchemaDefinition inputSchemaDefinition = null, SchemaDefinition outputSchemaDefinition = null)
+            where TSrc : class, new()
+            where TState : class, new()
+            where TDst : class, new()
+            => new StatefulCustomMappingEstimator<TSrc, TState, TDst>(catalog.GetEnvironment(), mapAction, contractName, stateInitAction, inputSchemaDefinition, outputSchemaDefinition);
+
+        /// <summary>
+        /// Drop rows where a specified predicate returns true.
+        /// </summary>
+        /// <typeparam name="TSrc">The class defining which columns to take from the incoming data.</typeparam>
+        /// <param name="catalog">The data operations catalog.</param>
+        /// <param name="input">The input data.</param>
+        /// <param name="filterPredicate">A predicate, that takes an input of type <typeparamref name="TSrc"/> and returns true if the row should be filtered (dropped) and false otherwise.</param>
+        /// <param name="inputSchemaDefinition">Additional parameters for schema mapping between <typeparamref name="TSrc"/> and input data. Useful
+        /// when dealing with annotations.</param>
+        public static IDataView CustomFilter<TSrc>(this DataOperationsCatalog catalog, IDataView input, Func<TSrc, bool> filterPredicate, SchemaDefinition inputSchemaDefinition = null)
+            where TSrc : class, new()
```

Comment:
Do we need samples illustrating these?

Reply:
Yes, good idea.

---
In reply to: [399050773](https://github.com/dotnet/machinelearning/pull/4676#discussion_r399050773) [](ancestors = 399050773)